Add Typing support for React Native 0.63

alexbrazier opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://reactnative.dev/docs/pressable https://reactnative.dev/docs/platformcolor
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

Basing off the original pr #44266 by @ph1p and adding additional types now that react native `0.63.0` has been released.

alexbrazier commented on types/react-native/index.d.ts:
```diff
@@ -7480,6 +7595,23 @@ export interface LinkingStatic extends NativeEventEmitter {
+     * Note: this only disables notifications, uncaught errors will still open a full screen LogBox.
+     * @param ignore whether to ignore logs or not
+     */
+    ignoreAllLogs(ignore: boolean): void;
```
For this function technically the original flow types made `ignore` optional and the code does `!!ignore` to make sure it's a bool. I decided to make it required because if you used `LogBox.ignoreAllLogs()` without a param it would default to `false` and enable logs, which is the opposite to what you would expect reading the code.

rickhanlonii replied:
Nice catch, fixed with https://github.com/facebook/react-native/pull/29310, we can update the type after that lands or we can land the types assuming it's fixed

alexbrazier replied:
@rickhanlonii Happy to do either, I guess it depends on when the release will be patched

alloy replied:
I agree with @rickhanlonii, regardless of when it’s released upstream.

rickhanlonii replied:
The change landed, not sure when the next pick and release to 0.63 will be

alexbrazier replied:
@alloy which of @rickhanlonii's options do you agree with? Changing to this now, or later?
```suggestion
    ignoreAllLogs(ignore?: boolean): void;
```

alloy replied:
Apologies, should’ve been more clear. I agree with **not** making it optional in the TS typings.

alexbrazier commented on types/react-native/index.d.ts:
```diff
@@ -5827,6 +5921,27 @@ interface PlatformWebStatic extends PlatformStatic {
+ *
+ * @see https://reactnative.dev/docs/platformcolor#example
+ */
+export function PlatformColor(color: string): string;
```
We might be able to add types to the `color` param instead of using a string to specify all the available ones for each platform.

This might get a bit complicated if the values change depending on the OS version so I just went with `string` for the initial version

TheSavior replied:
PlatformColor takes multiple arguments as fallbacks: https://github.com/facebook/react-native/blob/master/Libraries/StyleSheet/PlatformColorValueTypes.ios.js#L24-L26

TheSavior commented on types/react-native/index.d.ts:
```diff
@@ -5827,6 +5921,27 @@ interface PlatformWebStatic extends PlatformStatic {
+export function PlatformColor(color: string): string;
+
+type DynamicColorIOSTuple = {
+    light: string;
```
This should be ColorValue (at least, that's what it is in Flow, not sure what it is in these TypeScript definitions), as you can use any color, including a PlatformColor:

https://github.com/facebook/react-native/blob/ffa3d7f638c820dc208320193e6ba65667d751eb/Libraries/StyleSheet/PlatformColorValueTypes.ios.js#L28-L31

TheSavior commented on types/react-native/index.d.ts:
```diff
@@ -5827,6 +5921,27 @@ interface PlatformWebStatic extends PlatformStatic {
+ * @param tuple Colors you want to use for "light mode" and "dark mode"
+ * @platform ios
+ */
+export function DynamicColorIOS(tuple: DynamicColorIOSTuple): string;
```
@tom-un, can you please review the TypeScript types for the new colors methods? I know you were investigating ways to kinda align as best as possible with what we did in Flow relative to opaque types (which don't exist in TS, right?)

alexbrazier replied:
I ended up making `ColorValue` `any` for now because trying to do the actual recursive type was causing typescript errors with the current version being used. If I tried to updated it then other dependencies would break saying they specified a lower version.

I'm assuming having `any` is better than nothing as it means people can start using the new types now `0.63.0` has been released?

alexbrazier replied:
**Update:** I've created a separate PR with all the `ColorValue` changes that should make it easier to review https://github.com/alexbrazier/DefinitelyTyped/pull/1

@tom-un Would you be able to take a look when you get a chance?

tom-un replied:
Sorry I'm late to reviewing this.  I'm catching up.

I just backtracked the history and saw the three steps: first the `Opaque<K, T>` template approach, then the `Opaque<K>` approach with `K` a `unique symbol`, and then the final simplified form where the ColorValue is just `string | typeof OpaqueColorValue` with `OpaqueColorValue` a unique symbol.

I think I'm fine with the definition of the OpaqueColorValue.

But the latest iterations dropped the distinction between `ColorValue` and `ProcessedColorValue` that is in Flow.   The distinction in the Flow defs is important for internal component implementations.   I'm wondering if it would be important for an external component implementation that uses TypeScript and needs ColorValue props that could be semantic or a ProcessedColorValue.

TheSavior commented on types/react-native/test/index.tsx:
```diff
@@ -1086,6 +1114,13 @@ Platform.select({ android: 1, ios: 2, default: 0 }); // $ExpectType number
 Platform.select({ android: 1, ios: 2, macos: 3, web: 4, windows: 5 }); // $ExpectType number
 Platform.select({ android: 1, ios: 2, macos: 3, web: 4, windows: 5, default: 0 }); // $ExpectType number
 
+PlatformColor('?attr/colorControlNormal');
```
Please include a call with multiple fallback colors

TheSavior commented on types/react-native/test/index.tsx:
```diff
@@ -1086,6 +1114,13 @@ Platform.select({ android: 1, ios: 2, default: 0 }); // $ExpectType number
 
+PlatformColor('?attr/colorControlNormal');
+
+DynamicColorIOS({
+    dark: 'lightskyblue',
+    light: 'midnightblue',
+});
```
Can you add another one with a PlatformColor?
```
DynamicColorIOS({
    dark: 'lightskyblue',
    light: PlatformColor('labelColor'),
});
```

TheSavior commented on types/react-native/test/index.tsx:
```diff
@@ -356,6 +360,25 @@ export class TouchableNativeFeedbackTest extends React.Component {
+
+    render() {
+        return (
+            <Pressable onPress={this.onPressButton}>
```
As a test for Pressable, can you include the style/children as a function that takes press state approach since that is unique?

alexbrazier commented:
Thanks for the review @TheSavior, I've updated the PR with your suggestions.

TheSavior commented:
I'm deferring to Microsoft on the new Color types, as these types aren't correct. The new color methods don't return strings, they return an opaque type that should never ever be introspected or modified. We've been able to enforce that in Flow, I'm not sure how that can be enforced in TypeScript

alexbrazier commented:
@TheSavior I've opened up a second PR here with all the `ColorValue` changes. It is passing all the tests locally and seems to cover all the cases. Do you think this change is enough, and if so should I merge into this PR then get this PR merged?

I thought it would be easier to see the diff in a separate PR with only those changes.
https://github.com/alexbrazier/DefinitelyTyped/pull/1

Regarding the opaque types, because `ColorValue` is `null | string | someObjects...` and only `ColorValue` is exported, from my understanding this essentially achieves the same thing. You wouldn't be able to access the internals of the color without having to manually define a new type and explicitly cast it.

TheSavior commented:
That PR includes good changes, but doesn't address the opaque type issue.

The purpose of an opaque type in Flow is to ensure that the typechecker keeps people from writing styles like this:

```
<View style={{
  backgroundColor: {
      resource_paths: ['?attr/colorControlNormal']
  }
}} />
```
With the types as they are in Flow, you can't construct those objects manually, you can only create those objects by calling the helper functions PlatformColor() or DynamicColorIOS(). It also ensures that you don't call the helper function and then modify the resulting object. You can't do `PlatformColor('?attr/colorControlNormal').resource_paths.push('foo')` or any form of inspection or modification like that.

I'm not sure how to enforce this in TypeScript but I know @tom-un had thought about this previously.

For more info on opaque types, you can check out this blog post: https://codemix.com/opaque-types-in-javascript

alexbrazier commented:
Ah thanks, I didn't think of that case. I've now added an `Opaque` type and added the test cases mentioned above to make sure that you get an error.

As TypeScript doesn't have opaque types built in, it is possible to hack, but it would be pretty obvious when appending `__TYPE__: '__NativeColorValueAndroid__` to the type to try to set it.

I've updated the PR here https://github.com/alexbrazier/DefinitelyTyped/pull/1

alloy commented:
Regarding the `ColorValue`, if I understand this correctly, there’s really no need to expose _any_ details about the color, right? If so, a simple alternative would be to type it as a unique symbol:

```ts
declare const OpaqueColorValue: unique symbol;
type ColorValue = typeof OpaqueColorValue;
```

https://www.typescriptlang.org/play/index.html#code/CYUwxgNghgTiAEYD2A7AzgF3geQA5QEcBXEAYSQiRgDUoISAueIlAS2ITQE8BbAIwoBuAFAYuuBOUo06JeAF54YiUgBmOfBylVa9ECOGqWYDK1TwACtAyqqPbTAAUASiYPdcgN7D4v+HAwiGBR4TwBfeCg0eHdZfWEw4WFkdCwADwVLa1sYewoqF2E0gDo4NCQgsBAAfXwMAAtogHom+AASAFE0iRMOmBgqIA

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
     right?: number;
 }
 
+export interface PressableStateCallbackType {
```
Technically [the upstream type](https://github.com/facebook/react-native/blob/4f897336b6ce6c640a65f86b65c6d386ea55b944/Libraries/Components/Pressable/Pressable.js#L34) is read-only. I don’t feel very strong about this, but parity seems easy enough:

```suggestion
export type PressableStateCallbackType = Readonly<{
```

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Called when a single tap gesture is detected.
+     */
+    onPress?: (event: GestureResponderEvent) => void;
```
[The upstream type](https://github.com/facebook/react-native/blob/4f897336b6ce6c640a65f86b65c6d386ea55b944/Libraries/Components/Pressable/Pressable.js#L97) is a ‘maybe’ type, meaning it can be `null` too. Same applies to most (all?) other props below.

```suggestion
    onPress?: null | (event: GestureResponderEvent) => void;
```

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Called when a touch is engaged before `onPress`.
+     */
+    onPressIn?: (event: GestureResponderEvent) => void;
```
```suggestion
    onPressIn?: null | (event: GestureResponderEvent) => void;
```

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Called when a touch is released before `onPress`.
+     */
+    onPressOut?: (event: GestureResponderEvent) => void;
```
```suggestion
    onPressOut?: null | (event: GestureResponderEvent) => void;
```

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Duration (in milliseconds) from `onPressIn` before `onLongPress` is called.
+     */
+    delayLongPress?: number;
```
```suggestion
    delayLongPress?: null | number;
```

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Whether the press behavior is disabled.
+     */
+    disabled?: boolean;
```
```suggestion
    disabled?: null | boolean;
```

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Additional distance outside of this view in which a press is detected.
+     */
+    hitSlop?: Insets | number;
```
```suggestion
    hitSlop?: null | Insets | number;
```

alloy replied:
Might also be a good idea to [follow upstream](https://github.com/facebook/react-native/blob/4f897336b6ce6c640a65f86b65c6d386ea55b944/Libraries/StyleSheet/Rect.js#L20) and define a `RectOrSize` type.

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+     * Additional distance outside of this view in which a touch is considered a
+     * press before `onPressOut` is triggered.
+     */
+    pressRetentionOffset?: Insets | number;
```
```suggestion
    pressRetentionOffset?: null | Insets | number;
```

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * If true, doesn't play system sound on touch.
+     */
+    android_disableSound?: boolean;
```
```suggestion
    android_disableSound?: null | boolean;
```

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Enables the Android ripple effect and configures its color.
+     */
+    android_ripple?: RippleBackgroundPropType;
```
```suggestion
    android_ripple?: null | RippleBackgroundPropType;
```

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Used only for documentation or testing (e.g. snapshot testing).
+     */
+    testOnly_pressed?: boolean;
```
```suggestion
    testOnly_pressed?: null | boolean;
```

alloy commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+
+declare class PressableComponent extends React.Component<PressableProps> {}
+declare const PressableBase: Constructor<NativeMethodsMixinType> & typeof PressableComponent;
+export class Pressable extends PressableBase {}
```
Hmm, the upstream type is an ‘abstract component’ (for which we don’t have a equivalent yet), but I don’t believe it needs the native methods, no? (The component itself is also a function based one, not a class based one.)

alloy replied:
@alexbrazier This is my only remaining question—doesn’t seem like a blocking thing, though.

alexbrazier replied:
Ah, missed this one - any thoughts on what this would look like?

alloy commented on types/react-native/index.d.ts:
```diff
@@ -9341,7 +9470,7 @@ export const YellowBox: React.ComponentClass<any, any> & { ignoreWarnings: (warn
 /**
  * LogBox is enabled by default so there is no need to call unstable_enableLogBox() anymore. This is a no op and will be removed in the next version.
- * @depcreated
+ * @deprecated
```
👌

zoontek commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
 
+export type PressableStateCallbackType = Readonly<{
+    pressed: boolean;
+}
```
A missing `>` here

alloy replied:
My bad, thanks 🙏

zoontek commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Called when a single tap gesture is detected.
+     */
+    onPress?: null | (event: GestureResponderEvent) => void;
```
Missing parenthesis for all callbacks: `onPress?: null | ((event: GestureResponderEvent) => void);`

zoontek commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Called when a touch is engaged before `onPress`.
+     */
+    onPressIn?: null | (event: GestureResponderEvent) => void;
```
`onPressIn?: null | ((event: GestureResponderEvent) => void);`

zoontek commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Called when a touch is released before `onPress`.
+     */
+    onPressOut?: null | (event: GestureResponderEvent) => void;
```
`onPressOut?: null | ((event: GestureResponderEvent) => void);`

zoontek commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    /**
+     * Called when a long-tap gesture is detected.
+     */
+    onLongPress?: null | (event: GestureResponderEvent) => void;
```
`onLongPress?: null | ((event: GestureResponderEvent) => void);`

alexbrazier commented:
> Regarding the `ColorValue` type, if I understand this correctly, there’s really no need to expose _any_ details about the color, right? If so, a simple alternative would be to type it as a unique symbol:
> 
> ```ts
> declare const OpaqueColorValue: unique symbol;
> type ColorValue = typeof OpaqueColorValue;
> 
> function PlatformColor(): ColorValue {
>     return {} as ColorValue;
> }
> 
> const x = PlatformColor()
> x.resource_paths // $ExpectError
> ```
> 
> https://www.typescriptlang.org/play/index.html#code/CYUwxgNghgTiAEYD2A7AzgF3geQA5QEcBXEAYSQiRgDUoISAueIlAS2ITQE8BbAIwoBuAFAYuuBOUo06JeAF54YiUgBmOfBylVa9ECOGqWYDK1TwACtAyqqPbTAAUASiYPdcgN7D4v+HAwiGBR4TwBfeCg0eHdZfWEw4WFkdCwADwVLa1sYewoqF2E0gDo4NCQgsBAAfXwMAAtogHom+AASAFE0iRMOmBgqIA
> 
> I am not entirely opposed to the intersection solution you employed that’s described in [the aforementioned article](https://codemix.com/opaque-types-in-javascript/), especially as it may help debugging by including a self-describing label of what opaque type it is, but it should _not_ include the actual data structure typing as that makes it too easy to still manually construct such an object. E.g. this would be better imo:
> 
> ```ts
> type Opaque<K> = number & { __TYPE__: K };
> type NativeColorValueAndroid = Opaque<'__NativeColorValueAndroid__'>;
> ```
> 
> Do we really need different types for different platforms, though?

Thanks for the suggestions!. When I tried the first option I get errors that you cannot assign `string` to `backgroundColor` as the StyleSheet colors also use `ColorValue`.

I changed it to the second option which seems to have worked 🎉

TheSavior commented:
I think ColorValue would be a union of the unique symbol as well as string, just as it is in Flow:
```
declare const OpaqueColorValue: unique symbol;
type ColorValue = string | typeof OpaqueColorValue;

function PlatformColor(): OpaqueColorValue {
    return {} as ColorValue;
}
```

As you can use strings or the helpers to construct colors.

alexbrazier commented:
@TheSavior like this? https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45963/commits/f41faa4ab8f431107b305c619e94143ca0f64481

alloy commented:
@alexbrazier @TheSavior I wonder if technically the functions that produce platform colours should return only the opaque type, rather than a union of string and the opaque type? E.g.

```ts
declare const OpaqueColorValue: unique symbol;
type OpaqueColorValue = typeof OpaqueColorValue;
type ColorValue = string | OpaqueColorValue;

function DynamicColorIOS(tuple: DynamicColorIOSTuple): OpaqueColorValue;
function PlatformColor(...colors: string[]): OpaqueColorValue;
```

alexbrazier commented:
> @alexbrazier @TheSavior I wonder if technically the functions that produce platform colours should return only the opaque type, rather than a union of string and the opaque type? E.g.
> 
> ```ts
> declare const OpaqueColorValue: unique symbol;
> type OpaqueColorValue = typeof OpaqueColorValue;
> type ColorValue = string | OpaqueColorValue;
> 
> function DynamicColorIOS(tuple: DynamicColorIOSTuple): OpaqueColorValue;
> function PlatformColor(...colors: string[]): OpaqueColorValue;
> 
> const someColorProp: ColorValue = PlatformColor("test");
> ```

Yeah, I think that's better - https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45963/commits/35b6ea1d818850a167db2df005bb4acc25e240ac

I think that's what @TheSavior had proposed anyway, I had just misunderstood.

ph1p commented:
> Great stuff, thanks for the great work and bearing with us—same goes for @ph1p 👏 

Oh thank you ☺️

alexbrazier commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +471,84 @@ export interface Insets {
+    style?: StyleProp<ViewStyle> | ((state: PressableStateCallbackType) => StyleProp<ViewStyle>);
+}
+
+declare class PressableComponent extends React.Component<PressableProps> {}
+declare const PressableBase: Constructor<NativeMethodsMixinType> & typeof PressableComponent;
+export class Pressable extends PressableBase {}
```
```suggestion
export const Pressable: React.FunctionComponent<PressableProps>;
```

Prehaps just this?

alloy replied:
Yep, that seems fine to me, but leave a todo that it should really be an AbstractComponent 👍

alexbrazier replied:
Updated - https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45963/commits/e78b794c1b9553da2a88c68197febb8107da1c4b

tom-un commented:
> Hi @alexbrazier , @alloy, @TheSavior
> Regarding the `ColorValue` type (I commented elsewhere in the PR, but now cant find it):
> I'm concerned about the divergence from the Flow: in the Flow types there is a distinction between `ColorValue` and `ProcessedColorValue`. Previous iterations did have `ProcessedColorValue` but the latest does not.
> The distinction is critical in internal components, and for out-of-tree platforms (mac and windows) that will need to make their own opaque color value. But the distinction could also be important to an external component that implements using TypeScript instead of Flow.
> Also, it just would make the TypeScript parallel to the Flow defintions if we brought back a TS `ProcessedColorValue`.
> I think the current def of `ColorValue` is fine, but if we just had the additional def of `ProcessedColorValue` be `number | OpaqueColorValue`.
> And then the various props and arguments whose Flow types are `ColorValue | ProcessedColorValue` could be the same TypeScript type `ColorValue | ProcessedColorValue`.

Then the TypeScript signiture of `processColor` could change from
`export function processColor(color: any): number;`
to
`export function processColor(color?: ?(number | ColorValue)): ?ProcessedColorValue {


> Hi @alexbrazier , @alloy, @TheSavior
> Regarding the `ColorValue` type (I commented elsewhere in the PR, but now cant find it):
> I'm concerned about the divergence from the Flow: in the Flow types there is a distinction between `ColorValue` and `ProcessedColorValue`. Previous iterations did have `ProcessedColorValue` but the latest does not.
> The distinction is critical in internal components, and for out-of-tree platforms (mac and windows) that will need to make their own opaque color value. But the distinction could also be important to an external component that implements using TypeScript instead of Flow.
> Also, it just would make the TypeScript parallel to the Flow defintions if we brought back a TS `ProcessedColorValue`.
> I think the current def of `ColorValue` is fine, but if we just had the additional def of `ProcessedColorValue` be `number | OpaqueColorValue`.
> And then the various props and arguments whose Flow types are `ColorValue | ProcessedColorValue` could be the same TypeScript type `ColorValue | ProcessedColorValue`.

It looks like there are only a couple exported Flow instances of `ProcessedColorValue`: the return value of `processColor()` and the `tintColor` argument of `showActionSheetWithOptions()`.    But with a TypeScript form of `ProcessedColorValue` these two instances could be parallel with the Flow.   Hunting to see if there are more that I missed...

TheSavior commented:
Ah yeah, good call Tom! Definitely need `ProcessedColorValue` to be number | the opaque type to keep people from trying to introspect the results of `processColor` calls as it won't always return a number anymore.

alexbrazier commented:
Thanks @tom-un I've updated https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45963/commits/4d502564bb51acdd2e1d6802190c63aa9e707428

tom-un commented:
> Ah yeah, good call Tom! Definitely need `ProcessedColorValue` to be number | the opaque type to keep people from trying to introspect the results of `processColor` calls as it won't always return a number anymore.

I audited the rest of the exported color props and arguments.   I found one other minor case that is missing from the TypeScript altogether.   The PickerIOSItemProps in TS only has `label` and `value`.   But in the Flow equiv (RCTPickerIOSItemType) its actually `label`, `value`, and `textColor` where `textColor` is an optional `ProcessColorValue`. (see https://github.com/facebook/react-native/blob/8e6e83be0cf232aa299e64367fb223e4ce28d512/Libraries/Components/Picker/PickerIOS.ios.js#L41)

alexbrazier commented:
The tests are now failing due to `react-native-charts-wrapper` using `processColor` and expecting it to return a `number`. Should I update their types as well to use `ProcessedColorValue` instead of `number`?

tom-un commented on types/react-native/index.d.ts:
```diff
@@ -5827,6 +5910,34 @@ interface PlatformWebStatic extends PlatformStatic {
+
+export type ColorValue = string | OpaqueColorValue;
+
+type ProcessedColorValue = number | OpaqueColorValue;
```
I think this needs to have `export` because its used in other exported things

tom-un commented on types/react-native/index.d.ts:
```diff
@@ -9335,13 +9466,13 @@ export function findNodeHandle(
 ): null | number;
 
-export function processColor(color: any): number;
+export function processColor(color: any): ProcessedColorValue;
```
processColor can actually return ProcessedColorValue , undefined, or null.    also, the color argument shouldn't be any but instead an optional number | ColorValue.

The Flow decl looks like:
`function processColor(color?: ?(number | ColorValue)): ?ProcessedColorValue `

tom-un commented on types/react-native/index.d.ts:
```diff
@@ -3024,7 +3103,7 @@ export class PickerIOSItem extends React.Component<PickerIOSItemProps> {}
  */
```
Above the PickerIOSItemProps should have an optional `textColor` that is a `ProcessedColorValue` to be parallel with the Flow which is:

```
type RCTPickerIOSItemType = $ReadOnly<{|
  label: ?Label,
  value: ?(number | string),
  textColor: ?ProcessedColorValue,
|}>;
```

alexbrazier commented:
The build was failing because of `OpaqueColorValue` being `opaque symbol` instead of `symbol`. This caused an array of processed colors to have the type `symbol` when created using map, causing the build to fail. See this line https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-native-charts-wrapper/react-native-charts-wrapper-tests.tsx#L335

Updated in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45963/commits/9fe18bc38b4f64beb2353ccd304969afedd31c1e

Does it need to be a `unique symbol` or does `symbol` work?

If that's fine then it should be ready to go

alexbrazier commented:
@tom-un @alloy would you be able to take another look when you get the chance?

pesoklp13 commented on types/react-native/index.d.ts:
```diff
@@ -469,6 +472,83 @@ export interface Insets {
+    /**
+     * Called when a single tap gesture is detected.
+     */
+    onPress?: null | ((event: GestureResponderEvent) => void);
```
You can extract ((event: GestureResponderEvent) => void) into separate type.


```typescript
export type GestureResponseCallbackType = (event: GestureResponderEvent) => void;
```

```suggestion
    onPress?: null | GestureResponseCallbackType;
```

pesoklp13 commented on types/react-native-charts-wrapper/index.d.ts:
```diff
@@ -36,9 +36,9 @@ export type EasingType =
+ * use `processColor` from `react-native` to generate the corresponding number from a color (named, hex, rgba, etc.).
  */
-export type Color = number;
+export type Color = ProcessedColorValue | null | undefined;
```
Do you really need to use `undefined` here? I can see that every time it is used, it is something like `gridBackgroundColor?: Color;`

alexbrazier replied:
It's just becase the return type of `processColor` is `ProcessedColorValue | null | undefined` so now `Color` matches.

Also, the tests fail due to this line without `undefined` https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-native-charts-wrapper/react-native-charts-wrapper-tests.tsx#L402

alloy replied:
Perhaps it would be better to not explicitly define the type here, but rather get it from `processColor`?

E.g.

```suggestion
export type Color = ReturnType<typeof processColor>;
```

(Assuming `processColor` is imported in this module, otherwise that needs to be done too.)

alloy replied:
Added that while I was making edits anyways.

alloy commented:
> The build was failing because of `OpaqueColorValue` being `opaque symbol` instead of `symbol`. This caused an array of processed colors to have the type `symbol` when created using map, causing the build to fail. See this line https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-native-charts-wrapper/react-native-charts-wrapper-tests.tsx#L335
> 
> Updated in [9fe18bc](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/9fe18bc38b4f64beb2353ccd304969afedd31c1e)
> 
> Does it need to be a `unique symbol` or does `symbol` work?

It _has_ to be a unique symbol, otherwise any symbol would be assignable:

```ts
type X = symbol
let x: X = Symbol("foo")
x = Symbol("bar")
```

(I’ll take a look locally to see what the problem is and if I can figure out a solution.)

alloy commented:
For posterity sake, the error was:

```
Error: /home/runner/work/DefinitelyTyped/DefinitelyTyped/types/react-native-charts-wrapper/react-native-charts-wrapper-tests.tsx:324:25
358
ERROR: 324:25  expect  Compile error in typescript@3.3 but not in typescript@3.4.
359
Fix with a comment '// Minimum TypeScript Version: 3.4' just under the header.
360
Type '{ drawLabels: boolean; drawGridLines: boolean; position: string; yOffset: number; limitLines: { limit: number; lineColor: number | symbol | null | undefined; lineWidth: number; label: string; }[]; }' is not assignable to type 'xAxis'.
361
  Types of property 'position' are incompatible.
362
    Type 'string' is not assignable to type '"TOP" | "BOTTOM" | "BOTH_SIDED" | "TOP_INSIDE" | "BOTTOM_INSIDE" | undefined'.
```

And the [truncated] code that triggered it:

```ts
<CandleStickChart
    xAxis={{
        limitLines: Array.from({ length: 47 / 5 })
            .fill(0)
            .map((_, i) => {
                return {
                    lineColor: processColor('darkgray'),
                };
            }),
    }}
/>
```

From [the fix I just pushed](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45963/commits/b5cda864a9b54d6471741dc72a960757df0d8aa8):

> Because the array of limit lines is built dynamically, there is no way for older TS versions to know that the type of `lineColor` should remain the specific unique symbol, but is instead widened to any symbol.
>
> Explicitly typing the object literal fixes this, in similar way that literals can have their specificity preserved with `as const`.

For example:

```ts
<CandleStickChart
    xAxis={{
        limitLines: Array.from({ length: 47 / 5 })
            .fill(0)
            .map<{ lineColor: ProcessedColorValue }>((_, i) => {
                return {
                    lineColor: processColor('darkgray'),
                };
            }),
    }}
/>
```

alloy commented:
@tom-un I’ll wait for your go-ahead before merging.

alloy commented on types/react-native-charts-wrapper/index.d.ts:
```diff
@@ -82,6 +82,18 @@ export interface ChartLegend {
     };
 }
 
+export interface AxisLimitLine {
```
@dotconnor Any opinions on this name?

dotconnor replied:
LGTM 👍

alexbrazier commented:
Thanks for making the changes @alloy, they look good to me.

tom-un commented:
> @tom-un I’ll wait for your go-ahead before merging.

Looks great to me.   Thanks so much @alexbrazier and @alloy!

alloy commented:
> I would consider this https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45963/files#r453185295

@pesoklp13 I hear what you’re saying, but don’t consider it a blocker for a PR like this. I’m happy to merge a follow-up PR that cleans this up, but I also think it’s not a necessity.

alloy commented:
Great, let’s :shipit: then!
